src/utils/serial.cpp

```diff
@@ -1,20 +1,213 @@
+    LibFTDISerial (const char *description)
+        : description (description), port_open (false), logger (spdlog::get ("brainflow_logger"))
+    {
+    }
+
+    static bool is_libftdi (const char *port_name)
+    {
+        LibFTDISerial test (port_name);
+        return test.ctx.open (port_name) != -11;
+    }
+
+    void log_error (const char *action)
+    {
+        logger->error ("libftdi {}: {} -> {}", description, action, ctx.error_string ());
+    }
+
+    int open_serial_port ()
+    {
+        // https://www.intra2net.com/en/developer/libftdi/documentation/ftdi_8c.html#aae805b82251a61dae46b98345cd84d5c
+        switch (ctx.open (description))
+        {
+            case 0:
+                port_open = true;
+                return (int)SerialExitCodes::OK;
+            case -10:
+                log_error ("open_serial_port ()");
+                return (int)SerialExitCodes::CLOSE_ERROR;
+            default:
+                log_error ("open_serial_port ()");
+                return (int)SerialExitCodes::OPEN_PORT_ERROR;
+        }
+    }
+
+    bool is_port_open ()
+    {
+        return port_open;
+    }
+
+    int set_serial_port_settings (int ms_timeout = 1000, bool timeout_only = false)
+    {
+        int result;
+        if (!timeout_only)
+        {
+            result = ctx.set_line_property (BITS_8, STOP_BIT_1, NONE);
+            if (result != 0)
+            {
+                log_error ("set_serial_port_settings");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+            }
+            result = set_custom_baudrate (115200);
+            if (result != (int)SerialExitCodes::OK)
+            {
+                return result;
+            }
+            result = ctx.set_modem_control (Ftdi::Context::Dtr | Ftdi::Context::Rts);
+            result |= ctx.set_flow_control (SIO_DISABLE_FLOW_CTRL);
+            if (result != 0)
+            {
+                // -1 setting failed, -2 usb device unavailable
+                log_error ("set_serial_port_settings");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+            }
+        }
+
+        ctx.set_usb_read_timeout (ms_timeout);
+
+        return (int)SerialExitCodes::OK;
+    }
+
+    int set_custom_baudrate (int baudrate)
+    {
+        switch (ctx.set_baud_rate (115200))
+        {
+            case 0:
+                return (int)SerialExitCodes::OK;
+            case -3: // usb device unavailable
+                log_error ("set_custom_baudrate");
+                return (int)SerialExitCodes::OPEN_PORT_ERROR;
+            default: // -1 invalid baudrate, -2 setting baudrate failed
+                log_error ("set_custom_baudrate");
+                return (int)SerialExitCodes::SET_PORT_STATE_ERROR;
+        }
+    }
+
+    int flush_buffer ()
+    {
+        switch (ctx.tcflush (Ftdi::Context::Input | Ftdi::Context::Output))
+        {
+            case 0:
```

Comment:
are there enums or defines for such constants?

Reply:
They are actually specific to each function call, and specified in the api documentation per-function.  It's an old C library.  0 is always success.  Similar functions tend to have similar return codes.

We could of course make defines or enums for them.

Reply:
Here's the documentation for this particular function's return codes: https://www.intra2net.com/en/developer/libftdi/documentation/ftdi_8c.html#aeca5f2979d57490be0f3646fb4f3116a

Reply:
lets keep as is, no need to add own defines/enums